servicetalk-grpc-health/src/main/java/io/servicetalk/grpc/health/DefaultHealthService.java

```diff
@@ -0,0 +1,209 @@
+        // Try a get first to avoid locking with the assumption that most requests will be to watch existing services.
+        HealthValue healthValue = serviceToStatusMap.get(request.getService());
+        if (healthValue == null) {
+            if (!watchAllowed.test(request.getService())) {
+                return Publisher.failed(new GrpcStatus(FAILED_PRECONDITION, null, "watch not allowed for service " +
+                        request.getService()).asException());
+            }
+            lock.lock();
+            try {
+                if (terminated) {
+                    return Publisher.from(newBuilder().setStatus(NOT_SERVING).build());
+                }
+                healthValue = serviceToStatusMap.computeIfAbsent(request.getService(),
+                        __ -> new HealthValue(SERVICE_UNKNOWN));
+            } finally {
+                lock.unlock();
+            }
+        }
+
+        return Publisher.from(healthValue.last).concat(healthValue.publisher);
+    }
+
+    /**
+     * Updates the status of the server.
+     * @param service the name of some aspect of the server that is associated with a health status.
+     * This name can have no relation with the gRPC services that the server is running with.
+     * It can also be an empty String {@code ""} per the gRPC specification.
+     * @param status is one of the values {@link ServingStatus#SERVING}, {@link ServingStatus#NOT_SERVING},
+     * and {@link ServingStatus#UNKNOWN}.
+     * @return {@code true} if this change was applied. {@code false} if it was not due to {@link #terminate()}.
+     */
+    public boolean setStatus(String service, ServingStatus status) {
+        final HealthCheckResponse resp;
+        final HealthValue healthValue;
+        lock.lock();
+        try {
+            if (terminated) {
+                return false;
+            }
+            resp = newBuilder().setStatus(status).build();
+            healthValue = serviceToStatusMap.computeIfAbsent(service, __ -> new HealthValue(resp));
+        } finally {
+            lock.unlock();
+        }
+        healthValue.next(resp);
+        return true;
+    }
+
+    /**
+     * Clears the health status record of a service. The health service will respond with NOT_FOUND
+     * error on checking the status of a cleared service.
+     * @param service the name of some aspect of the server that is associated with a health status.
+     * This name can have no relation with the gRPC services that the server is running with.
+     * It can also be an empty String {@code ""} per the gRPC specification.
+     * @return {@code true} if this call removed a service. {@code false} if service wasn't found.
+     */
+    public boolean clearStatus(String service) {
+        final HealthValue healthValue = serviceToStatusMap.remove(service);
+        if (healthValue != null) {
+            healthValue.complete(SERVICE_UNKNOWN);
```

Comment:
Here is another problem now. If you do not clear the map on `terminated`, clearing status after termination will invoke `processor.onComplete()` twice. Our `PublisherProcessor` has protection agains that, but still not ideal to violate the spec here. If the `PublisherProcessor` impl changes later (or if we let users supply their own processor), it may cause an issue.
I don't think I 100% agreed with grpc-java impl to make this method noop after termination. It can still be useful to let users clear `NOT_SERVING` status.

Reply:
Yes our `PublisherProcessor` already protects against this. We could add additional protection here it currently isn't necessary and I would prefer to avoid adding complexity here. Let see start by adding comments and changing the method name.